### PR TITLE
Add doc about implementing an ONNX backend

### PR DESCRIPTION
**Pull request by bddppq**

### CODE REVIEW

File: docs/Implementing an ONNX backend.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+### Implementing an ONNX backend
+
+##### What is an ONNX backend
+
+An ONNX backend is a library that can run ONNX models. As there are already many existing Deep Learning frameworks, you very likely don't need to create everything from scratch but rather create a converter that converts ONNX models to the corresponding framework specific representation and then deligate the execution to the framework. E.g. [onnx-caffe2](https://github.com/onnx/onnx-caffe2) , [onnx-coreml](https://github.com/onnx/onnx-coreml), [onnx-tensorflow](https://github.com/onnx/onnx-tensorflow) are all implemented as converters.
```
Comment (marcelolr) [inline docs/Implementing an ONNX backend.md]: deligate->delegate

File: docs/Implementing an ONNX backend.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+##### Unified Backend Interface
+
+ONNX has defined an unified (Python) backend interface at https://github.com/onnx/onnx/blob/master/onnx/backend/base.py. There are three core concepts in this interface: `Device`, `Backend` and `BackendRep`.
```
Comment (marcelolr) [inline docs/Implementing an ONNX backend.md]: an unified->a unified

File: docs/Implementing an ONNX backend.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+- `Device` is a lightweight abstraction of varies hardwares, e.g. CPU, CUDA gpus.
```
Comment (marcelolr) [inline docs/Implementing an ONNX backend.md]: of varies hardwares->over varied hardware
CUDA gpus->GPUs (not sure it's worth clarifying CUDA-ness)

File: docs/Implementing an ONNX backend.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+- `BackendRep` is the handle that a Backend returns after prepared itself for executing a model repeatedly. Users will then pass inputs to the `run` function of `BackendRep` for getting the corresponding results.
+
+##### ONNX Backend Test
+
+ONNX has provided a standard backend test suite for helping backend implementations to do verification. We strongly encorage each ONNX backend to run it.
+
+There are two types of tests in this suite: Node Tests and Model Tests.
+
+- Node Tests is intended to verify whehter a backend is doing the correct computation, having the expected bahavior of handling varies attributes for each individual operator. In each test case, backend will be given a node and some inputs, its returned outputs will then be compared with the expected outputs.
```
Comment (marcelolr) [inline docs/Implementing an ONNX backend.md]: whether -> whether

**Comment by dzhulgakov** [comment]
Awesome!

Shall we also have a snippet on preferred way of adding tests, including model tests? Just code pointers are fine.

Also, maybe say that converter itself can be implemented in c++ and just have bindings?

**Pull request merged**